### DISCUSSION TITLE
Auth Exchange getAuth can't get local storage

### DISCUSSION THREAD

**Discussion by SleeplessOne1917** [urql-graphql/urql#2426] [Q&A] [answered]
I'm trying to use auth exchange in my app. On app startup, I fetch the token from local storage. My getAuth method is the following:
```
const getAuth = async ({ authState }) => {
	const token = localStorage.getItem('5etoken');

	if (!authState) {
		if (token) {
			return { token };
		}
		return null;
	}

	if (token) {
		const decoded = jwt.decode(token) as jwt.JwtPayload;

		if (decoded.exp !== undefined && decoded.exp < Date.now() / 1000) {
			return { token };
		}
	}

	return null;
};
```

When I start my app, I get an error saying localStorage is not defined.
The documentation for authentication uses local storage. The following is code from [the docs](https://formidable.com/open-source/urql/docs/advanced/authentication/#configuring-getauth-initial-load-fetch-from-storage) and I assume it works:
```
const getAuth = async ({ authState }) => {
  if (!authState) {
    const token = localStorage.getItem('token');
    const refreshToken = localStorage.getItem('refreshToken');
    if (token && refreshToken) {
      return { token, refreshToken };
    }
    return null;
  }

  return null;
};
```

While I could always check if `window` is defined before fetching from local storage, this makes it so I can never retrieve my token on app start up. Has anyone else run into this issue and had any luck solving it?

**Comment by SleeplessOne1917** [comment]
Looking into it more, it seems to have to do with Next JS doing server side rendering. If I manage to fix it I'll post what I did.

**Comment by SleeplessOne1917** [comment]
I'm completely stuck on this. I tried using Next JS's dynamic components with ssr set to false but that had no effect. I tried creating the client inside a useEffect hook, but that only made the provider use the default client. I tried using the higher order component from next-urql and only using the auth exchange when on the client, but it still ran on the server. I'm running out of ideas.

**Comment by JoviDeCroock** [comment, marked as answer]
@abias1122 

Well `localStorage` does not exist on the server so

```
const getAuth = async ({ authState }) => {
  if (typeof window === 'undefined') return null
  if (!authState) {
    const token = localStorage.getItem('token');
    const refreshToken = localStorage.getItem('refreshToken');
    if (token && refreshToken) {
      return { token, refreshToken };
    }
    return null;
  }

  return null;
};
```

Will make it work and authenticated requests won't be SSR'd which happens quite frequently, if you however want to SSR then you would have to use cookies as those are the only ones that are common between client and server.

That last part is more of a warning as this is more of a next and ssr related issue rather than urql 😅

**Reply by SleeplessOne1917** [reply]
I tried the approach you suggested and while it didn't cause me to get an error for `localStorage` being defined anymore, it has its own problems. If I return null if `window` is undefined, there's no way to set the initial auth state. This makes it so my app logs out on refresh. Is there any way to get around this?

**Reply by JoviDeCroock** [reply]
I mean, that's the second point I touched on 😅 authenticated are a choice you will need to make so you can for instance not do them on the server

**Reply by SleeplessOne1917** [reply]
I don't understand. I'm trying to avoid doing SSR for this particular case: ideally there's a way to make it so that `getAuth` only ever runs in the browser, even on startup. I may use your suggestion about cookies though, as it seems I'm forced into SSR regardless of my wishes.

**Reply by JoviDeCroock** [reply]
I mean, you can return null and just skip authenticated queries (paue: typeof window === 'undefined') as then you won't error out for auth-errors 😅 we can't skip getAuth as some people do auth on the server
